dotnet/src/SemanticKernel/Orchestration/Plan.cs

```diff
@@ -0,0 +1,281 @@
+    [JsonPropertyName("steps")]
+    internal List<Plan> Steps { get; } = new();
+
+    /// <summary>
+    /// Named parameters for the function
+    /// </summary>
+    [JsonPropertyName("named_parameters")]
+    public ContextVariables NamedParameters { get; set; } = new();
+
+    public bool HasNextStep => this.NextStep < this.Steps.Count;
+
+    public int NextStep { get; private set; } = 0;
+
+    #region ISKFunction implementation
+
+    /// <inheritdoc/>
+    [JsonPropertyName("name")]
+    public string Name { get; set; } = string.Empty;
+
+    /// <inheritdoc/>
+    [JsonPropertyName("skill_name")]
+    public string SkillName { get; set; } = string.Empty;
+
+    /// <inheritdoc/>
+    [JsonPropertyName("description")]
+    public string Description { get; set; } = string.Empty;
+
+    /// <inheritdoc/>
+    [JsonPropertyName("is_semantic")]
+    public bool IsSemantic { get; internal set; } = false;
+
+    /// <inheritdoc/>
+    [JsonPropertyName("request_settings")]
+    public CompleteRequestSettings RequestSettings { get; internal set; } = new();
+
+    #endregion ISKFunction implementation
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="Plan"/> class with a goal description.
+    /// </summary>
+    /// <param name="goal">The goal of the plan used as description.</param>
+    public Plan(string goal)
+    {
+        this.Description = goal;
+        this.SkillName = this.GetType().FullName;
+        this.Name = goal;
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="Plan"/> class with a goal description and steps.
+    /// </summary>
+    /// <param name="goal">The goal of the plan used as description.</param>
+    /// <param name="steps">The steps to add.</param>
+    public Plan(string goal, params ISKFunction[] steps) : this(goal)
+    {
+        this.AddStep(steps);
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="Plan"/> class with a goal description and steps.
+    /// </summary>
+    /// <param name="goal">The goal of the plan used as description.</param>
+    /// <param name="steps">The steps to add.</param>
+    public Plan(string goal, params Plan[] steps) : this(goal)
+    {
+        this.AddStep(steps);
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="Plan"/> class with a function.
+    /// </summary>
+    /// <param name="function">The function to execute.</param>
+    public Plan(ISKFunction function)
+    {
+        this.SetFunction(function);
+    }
+
+    /// <summary>
+    /// Adds one or more existing plans to the end of the current plan as steps.
+    /// </summary>
+    /// <param name="steps">The plans to add as steps to the current plan.</param>
+    /// <remarks>
+    /// When you add a plan as a step to the current plan, the steps of the added plan are executed after the steps of the current plan have completed.
+    /// </remarks>
+    public void AddStep(params Plan[] steps)
```

Comment:
Since a Plan is a ISKFunction, this method has no difference from the previous, I would remove this to avoid confusion or use the AddStep(ISK) impl as the method body: 

```
this.AddStep((ISKFunction[])steps);
```